src/libmongoc/src/mongoc/mongoc-server-monitor.c

```diff
@@ -0,0 +1,1241 @@
+{
+   bool ret = false;
+   mongoc_buffer_t buffer;
+   int32_t msg_len;
+   mongoc_rpc_t rpc;
+   bson_t reply_local;
+   int64_t expire_at_ms;
+   int64_t timeout_ms;
+
+   expire_at_ms = _now_ms () + server_monitor->heartbeat_frequency_ms +
+                  server_monitor->connect_timeout_ms;
+   _mongoc_buffer_init (&buffer, NULL, 0, NULL, NULL);
+   if (!_server_monitor_poll_with_interrupt (
+          server_monitor, expire_at_ms, cancelled, error)) {
+      GOTO (fail);
+   }
+
+   timeout_ms = BSON_MAX (expire_at_ms - _now_ms (), 1);
+   MONITOR_LOG (server_monitor,
+                "reading first 4 bytes with timeout: %" PRId64,
+                timeout_ms);
+   if (!_mongoc_buffer_append_from_stream (
+          &buffer, server_monitor->stream, 4, (int32_t) timeout_ms, error)) {
+      GOTO (fail);
+   }
+
+   BSON_ASSERT (buffer.len == 4);
+   memcpy (&msg_len, buffer.data, 4);
+   msg_len = BSON_UINT32_FROM_LE (msg_len);
+
+   if ((msg_len < 16) ||
+       (msg_len > server_monitor->description->max_msg_size)) {
+      bson_set_error (
+         error,
+         MONGOC_ERROR_PROTOCOL,
+         MONGOC_ERROR_PROTOCOL_INVALID_REPLY,
+         "Message size %d is not within expected range 16-%d bytes",
+         msg_len,
+         server_monitor->description->max_msg_size);
+      GOTO (fail);
+   }
+
+   timeout_ms = BSON_MAX (expire_at_ms - _now_ms (), 1);
+   MONITOR_LOG (server_monitor,
+                "reading remaining %d bytes. Timeout %" PRId64,
+                (int) (msg_len - 4),
+                timeout_ms);
+   if (!_mongoc_buffer_append_from_stream (
+          &buffer, server_monitor->stream, msg_len - 4, timeout_ms, error)) {
+      GOTO (fail);
+   }
+
+   if (!_mongoc_rpc_scatter (&rpc, buffer.data, buffer.len)) {
+      bson_set_error (error,
+                      MONGOC_ERROR_PROTOCOL,
+                      MONGOC_ERROR_PROTOCOL_INVALID_REPLY,
+                      "Malformed message from server");
+      GOTO (fail);
+   }
+
+   if (!_mongoc_rpc_decompress_if_necessary (&rpc, &buffer, error)) {
+      GOTO (fail);
+   }
+
+   _mongoc_rpc_swab_from_le (&rpc);
+   memcpy (&msg_len, rpc.msg.sections[0].payload.bson_document, 4);
+   msg_len = BSON_UINT32_FROM_LE (msg_len);
+   if (!bson_init_static (
+          &reply_local, rpc.msg.sections[0].payload.bson_document, msg_len)) {
+      bson_set_error (error,
+                      MONGOC_ERROR_PROTOCOL,
+                      MONGOC_ERROR_PROTOCOL_INVALID_REPLY,
+                      "Malformed BSON payload from server");
+      GOTO (fail);
+   }
+
+   bson_copy_to (&reply_local, ismaster_reply);
+   server_monitor->more_to_come =
+      (rpc.msg.flags & MONGOC_MSG_MORE_TO_COME) != 0;
+
+   ret = true;
+fail:
+   if (!ret) {
+      bson_init (ismaster_reply);
+   }
+   _mongoc_buffer_destroy (&buffer);
+   return ret;
+}
+
+/* Send and receive an awaitable ismaster.
+ *
+ * Called only from server monitor thread.
+ * Locks server monitor mutex.
```

Comment:
Similarly, `bson_mutex_lock` isn't called here. Is it called indirectly (perhaps when this defers to `_server_monitor_awaitable_ismaster_recv`)?